File: src/main/java/me/chriss99/worldmanagement/TileMap2D.java
package me.chriss99.worldmanagement;

import org.joml.Vector2i;

import java.util.*;
import java.util.function.BiConsumer;
import java.util.function.Function;

/**
 * Represents a map of Tiles, guarantees that ANY Vector2i can be asked for and stores previously loaded Tiles <br>
 * loading and unloading behaviour can be augmented with the TileLoadManager
 *
 * @param <T> Type of the Tiles to be managed
 */
public class TileMap2D<T> {
    protected final HashMap<Vector2i, T> loadedTiles = new LinkedHashMap<>();
    protected final Function<Vector2i, T> tileLoader;
    protected final BiConsumer<Vector2i, T> tileUnloader;
    protected TileLoadManager<T> loadManager;

    public TileMap2D(Function<Vector2i, T> tileLoader, BiConsumer<Vector2i, T> tileUnloader, TileLoadManager<T> loadManager) {
        this.tileLoader = tileLoader;
        this.tileUnloader = tileUnloader;
        this.loadManager = loadManager;
    }

    public T getTile(Vector2i coord) {
        return loadedTiles.computeIfAbsent(coord, tileLoader);
    }

    public Collection<T> getAllTiles() {
        return ((HashMap<Vector2i, T>) loadedTiles.clone()).values();
    }

    public void reloadAll() {
        loadedTiles.replaceAll((k, v) -> tileLoader.apply(k));
    }

    public void manageLoad() {
        HashSet<Vector2i> toRemove = new HashSet<>();

        loadedTiles.forEach((v, t) -> {
            if (!loadManager.loadPolicy(v, t))
                toRemove.add(v);
        });

        for (Vector2i v : loadManager.loadCommander())
            if (!toRemove.remove(v))
                getTile(v);

        for (Vector2i v : toRemove)
            tileUnloader.accept(v, loadedTiles.remove(v));
    }

    public void unloadAll() {
        loadedTiles.forEach(tileUnloader);
        loadedTiles.clear();
    }
}

File: src/main/kotlin/worldmanagement/TileMap2D.kt
package me.chriss99.worldmanagement

import glm_.vec2.Vec2i

/**
 * Represents a map of Tiles, guarantees that ANY Vec2i can be asked for and stores previously loaded Tiles <br></br>
 * loading and unloading behaviour can be augmented with the TileLoadManager
 *
 * @param <T> Type of the Tiles to be managed
</T> */
open class TileMap2D<T>(
    private val loadTile: (pos: Vec2i) -> T,
    private val unloadTile: (pos: Vec2i, tile: T) -> Unit,
    private var loadManager: TileLoadManager<T>
) {
    private val loadedTiles: HashMap<Vec2i, T> = LinkedHashMap<Vec2i, T>()
    val allTiles get() = loadedTiles.values.toList()

    operator fun get(coord: Vec2i): T = loadedTiles.computeIfAbsent(coord, loadTile)

    fun reloadAll() {
        loadedTiles.replaceAll { k: Vec2i, v: T ->
            unloadTile(k, v)
            loadTile(k)
        }
    }

    fun manageLoad() {
        val toRemove = HashSet<Vec2i>()

        loadedTiles.forEach { (v: Vec2i, t: T) ->
            if (!loadManager.loadPolicy(v, t))
                toRemove.add(v)
        }

        for (v in loadManager.loadCommander())
            if (!toRemove.remove(v))
                get(v)

        for (v in toRemove)
            unloadTile(v, loadedTiles.remove(v)!!)
    }

    fun unloadAll() {
        loadedTiles.forEach(unloadTile)
        loadedTiles.clear()
    }
}